### PR TITLE
Install mingw for ext compilation

### PR DESCRIPTION
**Pull request by CAMOBAP**
- https://github.com/lutaml/expressir/pull/33#issuecomment-748566905

### CODE REVIEW

**Comment by CAMOBAP** [comment]
@zakjan @ronaldtse 

Few moments to note:
 - Something definitely isn't right with the ubuntu compilation, a lot of unused args but most scared me is this one
    ```
   warning: type attributes ignored after type is already defined [-Wattributes]
   73 |   class ANTLR4CPP_PUBLIC ATN;
   ```
   This means that instead of `ATN` name class was named as `ANTLR4CPP_PUBLIC`, and it's a happy coincidence that this happened for just one class
 - rice gem compiles for windows but formally they [don't support it](https://github.com/jasonroelofs/rice/blob/master/README.md#installation-installation)
 - I'm not very familiar with ruby ext compilation, I just wondering why we don't use CMakeLists.txt which is provided by antlr library?

File: ext/express-parser/express_parser.cpp [hunk 1]
```diff
@@ -11,6 +11,15 @@
...
 
+#ifdef _WIN32
+// windef.h we will resists you!
+#undef FALSE
+#undef TRUE
+#undef OPTIONAL
+#undef IN
+#undef OUT
+#endif
+
```
Comment (zakjan) [inline ext/express-parser/express_parser.cpp:22]: This file is autogenerated with https://github.com/camertron/antlr4-native-rb. I make a few changes in it during the generation process in https://github.com/lutaml/expressir/blob/cpp-parser/exe/generate-parser#L23-L43, but I'd want to to send a PR upstream with them. I'm not sure if these Windows-specific lines shall be in the PR. Could they be moved to a separate file?

**Review by zakjan — APPROVED** [review]

File: ext/express-parser/express_parser.cpp (hunk 1 above)
Comment (CAMOBAP) [reply ext/express-parser/express_parser.cpp:22]: If those files autogenerated I think we should move them to the grammar file or to the script mentioned above. I will check this

**Comment by zakjan** [comment]
All C++ source files are generated either by ANTLR4 itself or https://github.com/camertron/antlr4-native-rb. Not sure why it doesn't use cmake. For Ruby native extension, I followed this guide https://guides.rubygems.org/gems-with-extensions/ Hopefully Rice is going to work on Windows, otherwise I'd need to create a custom wrapper instead.

Based on [ANTLR4 CPP target docs](https://github.com/antlr/antlr4/blob/master/doc/cpp-target.md#build-aspects), ANTLR4CPP_PUBLIC is used for specifying DLL export. Good catch, defining ANTLR4CPP_STATIC fixes it.

This can be merged into cpp-parser branch, than I'll update the generation script accordingly.

**Comment by CAMOBAP** [comment]
@zakjan thanks for the comment above those `#undef` also important to keep, because `windef.h` introduce defines which collides with ANTLR member-function names, this is why it's important to do this after rice's includes

**Comment by zakjan** [comment]
OK, I'll add it. I'm wondering why ANTLR4 docs doesn't mention this.

**Comment by CAMOBAP** [comment]
> I'm wondering why ANTLR4 docs doesn't mention this.

Because those includes comes from rice gem, I suppose. Its really hard to predict where legacy c++ will try to shoot you)

**Comment by zakjan** [comment]
All tests pass now 🎉

-Wattributes warnings are still present, but according to https://github.com/antlr/antlr4/issues/2450 they are fine.

**Comment by zakjan** [comment]
What's necessary to do for publishing a native extension to Rubygems?

**Comment by CAMOBAP** [comment]
@zakjan I'm working on a similar task for https://github.com/fontist/extract_ttc/issues/3 so you can leave this to me (just create a separate task for this and assign to me)

**Pull request merged**